valhalla/thor/bidirectional_astar.h

```diff
@@ -152,50 +154,294 @@ class BidirectionalAStar : public PathAlgorithm {
    * @param invariant          static date_time, dont offset the time as the path lengthens
    * @return returns true if the expansion continued from this node
    */
-  bool ExpandForward(baldr::GraphReader& graphreader,
-                     const baldr::GraphId& node,
-                     sif::BDEdgeLabel& pred,
-                     const uint32_t pred_idx,
-                     const baldr::TimeInfo& time_info,
-                     const bool invariant);
-  // Private helper function for `ExpandForward`
-  bool ExpandForwardInner(baldr::GraphReader& graphreader,
-                          const sif::BDEdgeLabel& pred,
-                          const baldr::NodeInfo* nodeinfo,
-                          const uint32_t pred_idx,
-                          const EdgeMetadata& meta,
-                          uint32_t& shortcuts,
-                          const graph_tile_ptr& tile,
-                          const baldr::TimeInfo& time_info);
+  template <const ExpansionType expansion_direction,
```

Comment:
Using a template here means (I think) that the compiler can simply "inject" the forward/reverse implementation of `Expand` right into the code.  Correct?

Another option would be to remove the templating and pass a new parameter into the routine, e.g., `expansion_direction` - which would be a runtime decision?

I can live with either approach, just bringing up some thoughts:
* If we did the "pass expansion_direction as an input argument", the overall amount of code in valhalla would be less.  (One implementation for `Expand` instead of two.), hence a smaller binary.
* The implementation for `Expand` could remain in a .cc file.  I know this is an open source library, but I still like to be mindful of "hiding the implementation" when possible.

Like I said, the template approach is cool, just thinking out loud.

Reply:
Over in https://github.com/valhalla/valhalla/pull/2969 a similar refactoring is occuring - I've moved the template implementation back into the `.cc` file to reduce clutter in the headers.

As this function is part of the core routing loop, I am concerned with performance - moving the conditional to a runtime parameter rather than a compile-time template parameter could risk some extra branching.  The *current* code has two separately named functions (`ExpandForward/ExpandReverse`) so the situation isn't getting any worse with this templatization, but the lines of source code is reduced :-)

I'm aiming for a zero-impact refactor for this change - no meaningful change in performance or binary sizes.  Once we start to consolidate the various algorithms that also duplicate these functions, we should start to see reductions in size simply by deleting duplicates.

Reply:
Sounds like a great plan, thanks.